src/Database/LSMTree/Monoidal.hs

```diff
@@ -56,34 +84,40 @@ class (MonadMVar m, MonadThrow m, MonadCatch m) => IOLike m where
 instance IOLike IO
 
 {-------------------------------------------------------------------------------
-  Table sessions
+  Sessions
 -------------------------------------------------------------------------------}
 
--- | Table context shared across multiple table handles, like counters and
--- filesystem information.
+-- | Context shared across multiple table handles, like counters and filesystem
+-- information.
 --
 -- For one, this is necessary if we want to be able to manipulate and query
 -- table handles, duplicate table handles, and load snapshots, all at the same
 -- time in the same directory.
-type TableSession :: (Type -> Type) -> Type -> Type -> Type
-data TableSession m k v = TableSession {
-    tsRoot   :: !FsPath
-  , tsConfig :: !TableConfig
-  , tsHasFS  :: !(SomeHasFS m)
+--
+-- Different types of tables can be live in the same session, but operations
+-- like 'merge' only work for __compatible__ tables: tables that belong to the
+-- same session, store the same key and value types, and have the same
+-- configuration parameters.
+type Session :: (Type -> Type) -> Type
+data Session m = Session {
+    sessionRoot  :: !FsPath
+  , sessionHasFS :: !(SomeHasFS m)
   }
 
--- | Table configuration parameters, like tuning parameters.
-data TableConfig
+-- | Check if two sessions are the same.
+--
+-- Necessary for checking table handle compatibility.
+isSameSession :: Session m -> Session m -> Bool
+isSameSession = undefined
```

Comment:
It's not totally clear that we need this. Yes the implementation of `merge` needs to check it, but it's not clear the caller needs to test it. The caller should be able to ensure it by construction without having to do a test to figure out which session is the right one. Same for the table config compatibility.

Reply:
I've removed it